File: Data Structures & Algorithms/partition-labels/submission-3.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        pos = {}
        for i, x in enumerate(s):
            pos[x] = i
        
        res = []
        size, end = 0, 0
        for i, c in enumerate(s):
            size += 1
            
            end = max(end, pos[c])

            if i == end:
                res.append(size)
                size = 0
        return res
            

        

File: Data Structures & Algorithms/non-overlapping-intervals/submission-3.java
class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> a[0] - b[0]);
        return intervals.length - dfs(0, -1, intervals);
    }

    private int dfs(int i, int prev, int[][] intervals) {
        if (i == intervals.length) {
            return 0;
        }
        //不选
        int res = dfs(i + 1, prev, intervals);
        //选
        if (intervals[i][0] >= prev || prev == -1) {
            res = Math.max(res, 1 + dfs(i + 1, intervals[i][1], intervals));
        }
        return res;
    }
}
